controllers/openshift_controller.go

```diff
@@ -1679,6 +1679,69 @@ func (r *KataConfigOpenShiftReconciler) getMcpByName(mcpName string) (*mcfgv1.Ma
 	return mcp, nil
 }
 
+func (r *KataConfigOpenShiftReconciler) processTransitioningNode(node *corev1.Node) error {
+	// Don't bother checking the second return value, the annotation is there
+	// otherwise we wouldn't be here (our caller checked already).
+	mcoNodeState, _ := node.Annotations["machineconfiguration.openshift.io/state"]
+
+	// Transitioning (Working & Degraded) nodes' handling only differs in
+	// which list a node ultimately goes to so we parameterise this.
+	var putOnInstallList, putOnUninstallList func(node *corev1.Node)
+
+	if mcoNodeState == "Working" {
+		putOnInstallList = func(node *corev1.Node) {
+			installInProgressList := &r.kataConfig.Status.InstallationStatus.InProgress.BinariesInstalledNodesList
+			*installInProgressList = append(*installInProgressList, node.GetName())
+		}
+		putOnUninstallList = func(node *corev1.Node) {
+			uninstallInProgressList := &r.kataConfig.Status.UnInstallationStatus.InProgress.BinariesUnInstalledNodesList
+			*uninstallInProgressList = append(*uninstallInProgressList, node.GetName())
+		}
+	} else { // mcoNodeState == "Degraded"
+		failedNodeStatus := kataconfigurationv1.FailedNodeStatus{
+			Name:  node.GetName(),
+			Error: node.Annotations["machineconfiguration.openshift.io/reason"],
+		}
+		putOnInstallList = func(node *corev1.Node) {
+			installFailedList := &r.kataConfig.Status.InstallationStatus.Failed.FailedNodesList
+			*installFailedList = append(*installFailedList, failedNodeStatus)
+		}
+		putOnUninstallList = func(node *corev1.Node) {
+			uninstallFailedList := &r.kataConfig.Status.UnInstallationStatus.Failed.FailedNodesList
+			*uninstallFailedList = append(*uninstallFailedList, failedNodeStatus)
+		}
+	}
+
```

Comment:
This `if {} else {}`  tends to indicate the code up to here rather belongs to some caller IMHO since it is presumably already branching based on the "machineconfiguration.openshift.io/state" annotation.

Maybe move these lines to intermediate `processTransitioningWorkingNode()` and `processTransitioningDegradedNode()` functions and keep the rest in `processTransitioningNode()` called by the other two ?

Reply:
Fair enough if you've tried it out before then ;-)